app/Tquery/Engine/TqBind.php

```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace App\Tquery\Engine;
+
+use App\Exceptions\FatalExceptionFactory;
+use Stringable;
+
+final readonly class TqBind implements Stringable
+{
+    private bool $used;
+
+    private function __construct(
+        public bool $isList,
+        public int $length,
+    ) {
+    }
+
+    public static function list(int $length): self
+    {
+        return new self(true, $length);
+    }
+
+    public static function single(): self
+    {
+        return new self(false, 1);
+    }
+
+    public function __toString(): string
+    {
+        // this class may support multiple __toString calls, then $builder->where $bindings
+        // must me multiplicated by number of calls: [1,2,3] => [1,2,3,1,2,3]
+        if ($this->used ?? false) {
+            FatalExceptionFactory::tquery()->throw();
```

Comment:
W sumie trochę nie rozumiem po co ten constraint - generalnie jak chcesz sobie wrzucić do query coś takiego 2. razy no to możesz (prawdopodobnie nie będzie to prawidłowe query, ale cóż. Generalnie obawiam się, że np ktoś będzie chciał sobie zalogować tego binda itp.

Ew. zrobić jakąś osobno nazwaną funkcję, jeśli Ci na tym ograniczeniu zależy - wtedy przynajmniej trzebą ja explicite zawołać i może mieć jakąś nazwę sugerującą, że obiekt jest potem do wyrzucenia.

Reply:
robienie zapytań `cośtam in $bind` wydaje mi się mniej zaciemniać, `cośtam in {$bind->tryToString()}` by komplikowało czytelność tego sql'a, który i tak zazwyczaj jest umiarkowanie czytelny

w ten sposób po prostu zabezpieczam, żeby ktoś nie wpisał `cośtam in $bind or coś_innego in $bind`, bo to stworzy błędne zapytanie. Zakładam, że błędy generowane przez bazę danych nie mogą się zdarzać przy poprawnym działaniu aplikacji. Przy czym mogę zrobić, że `cośtam in $bind or coś_innego in $bind` będzie poprawne, ale wtedy muszę wstawić wielokrotnie wartości do $bindings w $biulder->where

Reply:
zmieniłem na z `... $bind` na `... {$bind->use()}`, nadal mi się nie podoba :D